docs/design/language-support-notebook.md

```diff
@@ -0,0 +1,303 @@
+  - Maybe we want to debounce the `buildTarget/compile`
+
+## When VSCode request `textDocument/*`
+
+- We can convert notebook cell's URI `<->` generated Scala source for the cell, using `MappedSource`.
+- Conversion between position in the cell `<->` position in the generated Scala source can be done as the same way with Almond script.
+
+# Alternative Solution
+
+## Can we just combine cells into Ammonite script, instead of communicating with Almond BSP?
+
+While Almond can generate accurate Scala source program, it takes many steps for providing notebook support.
+
+Instead of using Almond, Metals can combine cells into `*.sc` file.
+
+For example, given the following notebook cells,
+
+```scala
+// cell1
+import $ivy.`io.circe::circe-generic:0.11.1`
+
+// cell2
+def add(a: Int, b: Int): Int = a + b
+
+// cell3
+add(1, 2)
+```
+
+Metals can combine those cells into `combined.sc`. The combine source will be a valid Ammmonite script file.
+Now Metals can re-use the Ammonite BSP for the language support.
+
+```scala
+import $ivy.`io.circe::circe-generic:0.11.1`
+def add(a: Int, b: Int): Int = a + b
+add(1, 2)
+```
+
+- pros
+  - We don't need to implement BSP for Almond, and re-use the Ammonite BSP.
+  - We will have `one notebook <-> one generated scala source`, that might be easier to handle (while Almond BSP will `one notebook cell <-> one generated scala source`).
+- cons
+  - It doesn't work if the program contains Almond-pre-imported things or [User APIs](https://almond.sh/docs/api).
+  - We have to run the same Scala version for Ammonite BSP with Almond Kernel, but it complicates the [Ammonite BSP logic](https://github.com/scalameta/metals/blob/fa03b00ad31307763979a00c4ccf794a2054c3fe/metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala#L184-L191) a bit.
+  - It adds up one more layer of conversion: `cell <-> combined sc file <-> Ammonite generated scala file`.
+
+## Run Almond BSP only when it's from VSCode
+
+Instead of running Almond BSP all the time, maybe we can run Almond BSP only when it's from VSCode?
+I don't think we can, because:
+
+- Looks like Kernel don't have information who is the clients, AFAIK.
+- Kernel may accept connection from multiple clients. Kernel started from VSCode doesn't mean it works only with Metals.
+
+## Almond BSP shouldn't accept `buildTarget/compile`?
+
+Instead of accepting `buildTarget/compile` and generate Scala sources on notebook changes, we can let Kernel compile Scala sources (enabling SemanticDB and writing out the generated Scala sources to the filesystem).
+
+It works only if Almond Kernel can detect it's running for Metals, but it looks like impossible as mentioned in the above section.
+
+# Milestones
+
+- [ ] Implement bare-minimum Almond BSP
+  - Create a sock file in a pre-defined location.
+  - Accept `build/initialize`
+  - `workspace/buildTargets` should returns the appropriate Scala version
```

Comment:
I would also include a request to execute/interpret a cell.
I don't know if we would be able to reuse existing method or it requires introduce a new one for BSP protocol.